fineract-provider/src/main/resources/META-INF/spring/securityContext.xml

```diff
@@ -42,19 +42,26 @@
 				method="POST" requires-channel="https" />
 			<intercept-url pattern="/api/*/self/registration/user" access="permitAll"
 				method="POST" requires-channel="https" />
-			<intercept-url pattern="/api/**" access="isFullyAuthenticated()"
+			<intercept-url pattern="/api/*/twofactor" access="isFullyAuthenticated()"
+						   method="GET" requires-channel="https" />
+			<intercept-url pattern="/api/*/twofactor" access="isFullyAuthenticated()"
+						   method="POST" requires-channel="https" />
+			<intercept-url pattern="/api/*/twofactor/validate" access="isFullyAuthenticated()"
+						   method="POST" requires-channel="https" />
+			<intercept-url pattern="/api/**" access="isFullyAuthenticated() and hasAuthority('TWOFACTOR_AUTHENTICATED')"
 				method="GET" requires-channel="https" />
-			<intercept-url pattern="/api/**" access="isFullyAuthenticated()"
+			<intercept-url pattern="/api/**" access="isFullyAuthenticated() and hasAuthority('TWOFACTOR_AUTHENTICATED')"
```

Comment:
In the implementation I am proposing, I look at 2fa as an authorisation problem. Users that are two-factor authenticated are granted the _TWOFACTOR_AUTHENTICATED_ authority. Endpoints under _/twofactor_ and _/twofactor/validate_ don't require fully authenticated uses, thus no need to check for TWOFACTOR_AUTHENTICATED.